File: persistence/src/main/java/io/github/manami/persistence/importer/xml/parser/ManamiSaxParser.java
package io.github.manami.persistence.importer.xml.parser;

import static com.google.common.collect.Lists.newArrayList;

import io.github.manami.dto.AnimeType;
import io.github.manami.dto.ToolVersion;
import io.github.manami.dto.entities.Anime;
import io.github.manami.dto.entities.FilterEntry;
import io.github.manami.dto.entities.InfoLink;
import io.github.manami.dto.entities.WatchListEntry;
import io.github.manami.persistence.PersistenceFacade;
import io.github.manami.persistence.importer.xml.postprocessor.ImportMigrationPostProcessor;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.List;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.xml.sax.Attributes;
import org.xml.sax.SAXException;
import org.xml.sax.helpers.DefaultHandler;

/**
 * @author manami-project
 * @version 2.0.0
 */
public class ManamiSaxParser extends DefaultHandler {

  private static final Logger log = LoggerFactory.getLogger(ManamiSaxParser.class);
  private static final String THUMBNAIL_PARSING_EXCEPTION = "Unable to parse thumbnail URL from [{}]";

  /**
   * This is the builder for the text within the elements.
   */
  private StringBuilder strBuilder;

  private final PersistenceFacade persistence;

  private final List<Anime> animeListEntries;
  private final List<FilterEntry> filterListEntries;
  private final List<WatchListEntry> watchListEntries;

  private ImportMigrationPostProcessor importMigrationPostProcessor;


  /**
   * Constructor awaiting a list.
   *
   * @version 2.0.0
   */
  public ManamiSaxParser(final PersistenceFacade persistence) {
    this.persistence = persistence;
    animeListEntries = newArrayList();
    filterListEntries = newArrayList();
    watchListEntries = newArrayList();
  }


  @Override
  public void startElement(final String namespaceURI, final String localName, final String qName, final Attributes attributes) {
    strBuilder = new StringBuilder();

    switch (qName) {
      case "manami":
        importMigrationPostProcessor = new ImportMigrationPostProcessor(ToolVersion.getToolVersion(),
            attributes.getValue("version"),
            animeListEntries,
            filterListEntries,
            watchListEntries);
        break;
      case "anime":
        createAnimeEntry(attributes);
        break;
      case "filterEntry":
        createFilterEntry(attributes);
        break;
      case "watchListEntry":
        createWatchListEntry(attributes);
        break;
    }
  }


  @Override
  public void endDocument() throws SAXException {
    persistence.addAnimeList(animeListEntries);
    persistence.addFilterList(filterListEntries);
    persistence.addWatchList(watchListEntries);
    importMigrationPostProcessor.execute();
  }


  /**
   * @since 2.7.0
   */
  private void createAnimeEntry(final Attributes attributes) {
    final String title = attributes.getValue("title").trim();
    final InfoLink infoLink = new InfoLink(attributes.getValue("infoLink").trim());

    final Anime actAnime = new Anime(title, infoLink);
    actAnime.setType(AnimeType.findByName(attributes.getValue("type").trim()));
    actAnime.setEpisodes(Integer.valueOf(attributes.getValue("episodes").trim()));
    actAnime.setLocation(attributes.getValue("location").trim());

    animeListEntries.add(actAnime);
  }


  /**
   * @since 2.7.0
   */
  private void createFilterEntry(final Attributes attributes) {
    try {
      final FilterEntry entry = new FilterEntry(attributes.getValue("title").trim(),
          new InfoLink(attributes.getValue("infoLink").trim()),
          new URL(attributes.getValue("thumbnail").trim()));

      filterListEntries.add(entry);
    } catch (MalformedURLException e) {
      log.error(THUMBNAIL_PARSING_EXCEPTION, attributes.getValue("title"));
    }
  }


  /**
   * @since 2.8.0
   */
  private void createWatchListEntry(final Attributes attributes) {
    try {
      final WatchListEntry entry = new WatchListEntry(attributes.getValue("title").trim(),
          new InfoLink(attributes.getValue("infoLink").trim()),
          new URL(attributes.getValue("thumbnail").trim()));

      watchListEntries.add(entry);
    } catch (MalformedURLException e) {
      log.error(THUMBNAIL_PARSING_EXCEPTION, attributes.getValue("title"));
    }
  }


  @Override
  public void characters(final char ch[], final int start, final int length) {
    strBuilder.append(new String(ch, start, length));
  }
}

File: persistence/src/main/kotlin/io/github/manami/persistence/importer/xml/parser/ManamiSaxParser.kt
package io.github.manami.persistence.importer.xml.parser

import io.github.manami.common.LoggerDelegate
import io.github.manami.dto.AnimeType
import io.github.manami.dto.entities.Anime
import io.github.manami.dto.entities.FilterListEntry
import io.github.manami.dto.entities.InfoLink
import io.github.manami.dto.entities.WatchListEntry
import io.github.manami.persistence.InternalPersistence
import io.github.manami.persistence.importer.xml.postprocessor.ImportDocument
import io.github.manami.persistence.importer.xml.postprocessor.ImportMigrationPostProcessor
import org.slf4j.Logger
import org.xml.sax.Attributes
import org.xml.sax.helpers.DefaultHandler
import java.lang.StringBuilder
import java.net.MalformedURLException
import java.net.URL


private const val THUMBNAIL_PARSING_EXCEPTION: String = "Unable to parse thumbnail URL from [{}]"


internal class ManamiSaxParser(
        private val persistence: InternalPersistence
) : DefaultHandler() {

    private val log: Logger by LoggerDelegate()

    /**
     * This is the builder for the text within the elements.
     */
    private var strBuilder = StringBuilder()
    private val importDocument = ImportDocument(
            "",
            mutableListOf(),
            mutableListOf(),
            mutableListOf()
    )

    override fun startElement(namespaceUri: String, localName: String, qName: String, attributes: Attributes) {
        strBuilder = StringBuilder()

        when (qName) {
            "manami" -> importDocument.documentVersion = attributes.getValue("version")
            "anime" -> createAnimeEntry(attributes)
            "filterEntry" -> createFilterEntry(attributes)
            "watchListEntry" -> createWatchListEntry(attributes)
        }
    }


    override fun endDocument() {
        persistence.addAnimeList(importDocument.animeListEntries)
        persistence.addFilterList(importDocument.filterListEntries)
        persistence.addWatchList(importDocument.watchListEntries)
        ImportMigrationPostProcessor.process(importDocument)
    }


    private fun createAnimeEntry(attributes: Attributes) {
        val title: String = attributes.getValue("title").trim()
        val infoLink = InfoLink(attributes.getValue("infoLink").trim())

        val actAnime: Anime = Anime(title, infoLink).apply {
            episodes = Integer.valueOf(attributes.getValue("episodes").trim())
            location = attributes.getValue("location").trim()
        }

        AnimeType.findByName(attributes.getValue("type").trim())?.let {
            actAnime.type = it
        }

        importDocument.animeListEntries.add(actAnime)
    }


    private fun createFilterEntry(attributes: Attributes) {
        try {
            val entry = FilterListEntry(
                    attributes.getValue("title").trim(),
                    InfoLink(attributes.getValue("infoLink").trim()),
                    URL(attributes.getValue("thumbnail").trim())
            )

            importDocument.filterListEntries.add(entry)
        } catch (e: MalformedURLException) {
            log.error(THUMBNAIL_PARSING_EXCEPTION, attributes.getValue("title"))
        }
    }


    private fun createWatchListEntry(attributes: Attributes) {
        try {
            val entry = WatchListEntry(
                    attributes.getValue("title").trim(),
                    InfoLink(attributes.getValue("infoLink").trim()),
                    URL(attributes.getValue("thumbnail").trim())
            )

            importDocument.watchListEntries.add(entry)
        } catch (e: MalformedURLException) {
            log.error(THUMBNAIL_PARSING_EXCEPTION, attributes.getValue("title"))
        }
    }


    override fun characters(ch: CharArray, start: Int, length: Int) {
        strBuilder.append(String(ch, start, length))
    }
}